packages/cli/command_parser.pony

```diff
@@ -0,0 +1,340 @@
+        | let se: SyntaxError => return se
+        end
+      | None => SyntaxError(short_string(c), "unknown short option")
+      else
+        // TODO: Ponyc should know we've covered all match cases above
+        return SyntaxError(token, "Pony shouldn't allow this")
+      end
+    end
+    options
+
+  fun box _parse_arg(token: String, arg_pos: USize): (Arg | SyntaxError) =>
+    try
+      let arg_spec = _spec.args(arg_pos)
+      ArgParser.parse(arg_spec, token)
+    else
+      return SyntaxError(token, "too many positional arguments")
+    end
+
+  fun box _option_with_name(name: String): (OptionSpec | None) =>
+    try
+      return _spec.options(name)
+    end
+    match _parent
+    | let p: CommandParser box => p._option_with_name(name)
+    else
+      None
+    end
+
+  fun box _option_with_short(short: U8): (OptionSpec | None) =>
+    for f in _spec.options.values() do
+      if f._has_short(short) then
+        return f
+      end
+    end
+    match _parent
+    | let p: CommandParser box => p._option_with_short(short)
+    else
+      None
+    end
+
+  fun short_string(c: U8): String =>
+    recover String.from_utf32(c.u32()) end
+
+  fun box _help_name(): String =>
+    _root_spec().help_name
+
+
+primitive OptionParser
+  fun parse(
+    spec: OptionSpec,
+    targ: (String|None),
+    args: Array[String] ref): (Option | SyntaxError)
+  =>
+    // Grab the token-arg if provided, else consume an arg if one is required.
+    let arg = match targ
+      | (let fn: None) if spec._requires_arg() =>
+        try args.shift() else None end
+      else
+        targ
+      end
+    // Now convert the arg to Type, detecting missing or mis-typed args
+    match arg
+    | let a: String =>
+      match ValueParser.parse(spec.typ, a)
+      | let v: Value => Option(spec, v)
+      | let se: SyntaxError => se
+      else
+          // TODO: Ponyc should know we've covered all match cases above
+          SyntaxError(a, "Pony shouldn't allow this")
+      end
+    else
+      if not spec._requires_arg() then
+        Option(spec, spec._default_arg())
+      else
+        SyntaxError(spec.name, "missing arg for option")
+      end
+    end
+
+
+primitive ArgParser
+  fun parse(spec: ArgSpec, arg: String): (Arg | SyntaxError) =>
+    match ValueParser.parse(spec.typ, arg)
+    | let v: Value => Arg(spec, v)
+    | let se: SyntaxError => se
+    else
+        // TODO: Ponyc should know we've covered all match cases above
+        SyntaxError(arg, "Pony shouldn't allow this")
+    end
+
+
+primitive ValueParser
+  fun box parse(typ: ValueType, arg: String): (Value | SyntaxError) =>
+    try
+      match typ
+      | let b: BoolType => arg.bool()
+      | let s: StringType => arg
+      | let f: F64Type => arg.f64()
+      | let i: I64Type => arg.i64()
+      else
+        // TODO: Ponyc should know we've covered all match cases above
+        SyntaxError(arg,
+          "Pony shouldn't allow this: unknown value type " + typ.string())
```

Comment:
This will be fixed when my "exhaustive match" PR (#1891) gets merged, which will hopefully happen today. So you might want to subscribe to that PR, and update this PR by removing the unnecessary "else" clause here if my PR is merged first.